ugs-platform/ugs-platform-plugin-setup-wizard/src/main/java/com/willwinder/ugs/nbp/setupwizard/panels/WizardPanelHardLimits.java

```diff
@@ -0,0 +1,126 @@
+    private JLabel labelExtendedDescription;
+
+    public WizardPanelHardLimits(BackendAPI backend) {
+        super(backend, "Limit switches");
+
+        initComponents();
+        initLayout();
+        setValid(true);
+    }
+
+    private void initLayout() {
+        JPanel panel = new JPanel(new MigLayout("fillx, inset 0, gap 5, hidemode 3"));
+        panel.add(labelDescription, "growx, wrap, gapbottom 10");
+        panel.add(labelExtendedDescription, "wrap, gapbottom 10");
+        panel.add(checkboxEnableHardLimits, "wrap");
+        panel.add(labelHardLimitssNotSupported, "wrap");
+        add(panel);
+    }
+
+    private void initComponents() {
+        labelDescription = new JLabel("<html><body>" +
+                "<p>Limit switches will prevent the machine to move beyond its physical limits.</p>" +
+                "</body></html>");
+
+        labelExtendedDescription = new JLabel("<html><body>" +
+                "<p>Before enabling make sure that you have equipped your machine with switches on at least one end for each axis.</p>" +
+                "</body></html>");
+
+        checkboxEnableHardLimits = new JCheckBox("Enable limit switches");
+        checkboxEnableHardLimits.addActionListener(event -> onHardLimitsClicked());
+
+        labelHardLimitssNotSupported = new JLabel("<html><body>Limit switches are unfortunately not available on your hardware.</body></html>", ImageUtilities.loadImageIcon("icons/information24.png", false), JLabel.LEFT);
+        labelHardLimitssNotSupported.setVisible(false);
+    }
+
+    private void onHardLimitsClicked() {
+        try {
+            getBackend().getController().getFirmwareSettings().setHardLimitsEnabled(checkboxEnableHardLimits.isSelected());
+        } catch (FirmwareSettingsException e) {
+            NotifyDescriptor nd = new NotifyDescriptor.Message("Couldn't enable/disable the hard limits settings: " + e.getMessage(), NotifyDescriptor.ERROR_MESSAGE);
+            DialogDisplayer.getDefault().notify(nd);
+        }
+    }
+
+    @Override
+    public void initialize() {
+        getBackend().addUGSEventListener(this);
+        if (getBackend().getController() != null && getBackend().getController().getCapabilities().hasHardLimits()) {
+            IFirmwareSettings firmwareSettings = getBackend().getController().getFirmwareSettings();
+            checkboxEnableHardLimits.setSelected(firmwareSettings.isHardLimitsEnabled());
+            checkboxEnableHardLimits.setVisible(true);
+            labelExtendedDescription.setVisible(true);
+            labelHardLimitssNotSupported.setVisible(false);
+        } else {
+            labelExtendedDescription.setVisible(false);
+            checkboxEnableHardLimits.setVisible(false);
+            labelHardLimitssNotSupported.setVisible(true);
+        }
+    }
+
+    @Override
+    public void destroy() {
+        getBackend().removeUGSEventListener(this);
+    }
+
+    @Override
+    public boolean isEnabled() {
```

Comment:
Each step is now responsible for letting the wizard know if the step is applicable for the controller.